src/Analyzer/Passes/CNF.cpp

```diff
@@ -0,0 +1,592 @@
+#include <Analyzer/Passes/CNF.h>
+
+#include <Analyzer/InDepthQueryTreeVisitor.h>
+#include <Analyzer/FunctionNode.h>
+#include <Analyzer/ConstantNode.h>
+
+#include <Interpreters/TreeCNFConverter.h>
+
+#include <IO/WriteBufferFromString.h>
+#include <IO/Operators.h>
+
+#include <Functions/FunctionFactory.h>
+
+#include <Common/checkStackSize.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int TOO_MANY_TEMPORARY_COLUMNS;
+}
+
+namespace Analyzer
+{
+
+namespace
+{
+
+bool isLogicalFunction(const FunctionNode & function_node)
+{
+    const std::string_view name = function_node.getFunctionName();
+    return name == "and" || name == "or" || name == "not";
+}
+
+template <typename... Args>
+QueryTreeNodePtr createFunctionNode(const FunctionOverloadResolverPtr & function_resolver, Args &&... args)
+{
+    auto function_node = std::make_shared<FunctionNode>(function_resolver->getName());
+    auto & new_arguments = function_node->getArguments().getNodes();
+    new_arguments.reserve(sizeof...(args));
+    (new_arguments.push_back(std::forward<Args>(args)), ...);
+    function_node->resolveAsFunction(function_resolver);
+    return function_node;
+}
+
+size_t countAtoms(const QueryTreeNodePtr & node)
+{
+    checkStackSize();
+
+    const auto * function_node = node->as<FunctionNode>();
+    if (!function_node || !isLogicalFunction(*function_node))
+        return 1;
+
+    size_t atom_count = 0;
+    const auto & arguments = function_node->getArguments().getNodes();
+    for (const auto & argument : arguments)
+        atom_count += countAtoms(argument);
+
+    return atom_count;
+}
+
+class SplitMultiLogicVisitor
+{
+public:
+    explicit SplitMultiLogicVisitor(ContextPtr context)
+        : current_context(std::move(context))
+    {}
+
+    void visit(QueryTreeNodePtr & node)
+    {
+        checkStackSize();
+
+        auto * function_node = node->as<FunctionNode>();
+        if (!function_node || !isLogicalFunction(*function_node))
+            return;
+
+        const auto & name = function_node->getFunctionName();
+
+        if (name == "and" || name == "or")
+        {
+            auto function_resolver = FunctionFactory::instance().get(name, current_context);
+
+            const auto & arguments = function_node->getArguments().getNodes();
+            if (arguments.size() > 2)
+            {
+                QueryTreeNodePtr current = arguments[0];
+                for (size_t i = 1; i < arguments.size(); ++i)
+                    current = createFunctionNode(function_resolver, std::move(current), arguments[i]);
+
+                auto & new_function_node = current->as<FunctionNode &>();
+                function_node->getArguments().getNodes() = std::move(new_function_node.getArguments().getNodes());
+                function_node->resolveAsFunction(function_resolver);
```

Comment:
```
Definition of implicit copy assignment operator for 'IQueryTreeNode' is deprecated because it has a user-declared destructor (clang -Wdeprecated-copy-with-dtor)
```